File: voronoi/integration.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <numeric>
#include <tuple>
#include <Eigen/Dense>
#include <boost/math/quadrature/gauss_kronrod.hpp>
#include <boost/math/constants/constants.hpp>
// 1) integrate voro++ so that we can obtain vertices of the polyhedrons starting from robot pose
// 2) publish control input to the robot
using namespace std;
using namespace Eigen;
namespace bmq = boost::math::quadrature;

// Define multivariate Gaussian PDF parameters
VectorXd mean(3); // Mean vector
MatrixXd covariance(3, 3); // Covariance matrix

// Multivariate Gaussian PDF in R^3
VectorXd multivariate_gaussian_pdf(double x, double y, double z) {
    Vector3d point(x, y, z);
    Vector3d diff = point - mean;
    double exponent = -0.5 * diff.transpose() * covariance.inverse() * diff;
    double normalization = pow(2 * boost::math::constants::pi<double>(), -mean.size() / 2) * sqrt(covariance.determinant());
    VectorXd density(3);
    density << normalization * exp(exponent), 0, 0; // Return a vector with the density for each component (in this case, just one component)
    return density;
}

// Product of [x, y, z] and the result of multivariate_gaussian_pdf(x, y, z)
VectorXd product_pdf(double x, double y, double z) {
    VectorXd pdf = multivariate_gaussian_pdf(x, y, z);
    VectorXd result = VectorXd::Zero(3);
    result << x * pdf(0), y * pdf(0), z * pdf(0); // Element-wise product with [x, y, z]
    return result;
}

// Define polyhedron volume
double polyhedron_volume(vector<Vector3d>& vertices) {
    MatrixXd mat(3, vertices.size());
    for (size_t i = 0; i < vertices.size(); ++i) {
        mat.col(i) = vertices[i];
    }
    return mat.jacobiSvd().singularValues().prod() / 6.0; // Using SVD to compute the volume of a tetrahedron
}

// Integrate vector-valued PDF over the polyhedron
VectorXd integrate_vector_valued_pdf_over_polyhedron(std::function<VectorXd(double, double, double)> pdf_func, vector<Vector3d>& vertices) {
    // Compute the volume of the polyhedron
    double polyhedron_vol = polyhedron_volume(vertices);

    // Define the integration limits
    double xmin = numeric_limits<double>::infinity(), xmax = -numeric_limits<double>::infinity();
    double ymin = numeric_limits<double>::infinity(), ymax = -numeric_limits<double>::infinity();
    double zmin = numeric_limits<double>::infinity(), zmax = -numeric_limits<double>::infinity();
    for (const auto& vertex : vertices) {
        xmin = min(xmin, vertex(0));
        xmax = max(xmax, vertex(0));
        ymin = min(ymin, vertex(1));
        ymax = max(ymax, vertex(1));
        zmin = min(zmin, vertex(2));
        zmax = max(zmax, vertex(2));
    }

    // Perform the triple integral over the polyhedron for each component separately
    VectorXd result(3);
    for (int i = 0; i < 3; ++i) {
        auto integrator_func = [&](double z) {
            auto integrator_func_y = [&](double y) {
                auto integrator_func_x = [&](double x) {
                    return pdf_func(x, y, z)(i); // Extract the ith component
                };
                return bmq::gauss_kronrod<double, 15>().integrate(integrator_func_x, xmin, xmax);
            };
            return bmq::gauss_kronrod<double, 15>().integrate(integrator_func_y, ymin, ymax);
        };
        result(i) = bmq::gauss_kronrod<double, 15>().integrate(integrator_func, zmin, zmax);
    }

    // Multiply each component by the volume of the polyhedron
    return result * polyhedron_vol;
}

int main() {
    // Set mean and covariance of the multivariate Gaussian PDF
    mean << 0, 0, 0; // Mean vector
    covariance = MatrixXd::Identity(3, 3); // Identity covariance matrix

    // Example usage: define the vertices of a tetrahedron
    vector<Vector3d> vertices = {Vector3d(0, 0, 0), Vector3d(1, 0, 0), Vector3d(0, 1, 0), Vector3d(0, 0, 1)};
    
    // Integrate the product of vector-valued PDF over the polyhedron
    VectorXd result = integrate_vector_valued_pdf_over_polyhedron(product_pdf, vertices);
    cout << "Result: " << result.transpose() << endl;

    return 0;
}

File: scripts/integration.py
import numpy as np
from scipy.spatial import ConvexHull
from scipy.integrate import quad


X_LOW = -0.5
X_HIGH = 0.5
Y_LOW = -0.5
Y_HIGH = 0.5
Z_LOW = -0.5
Z_HIGH = 0.5

# # Define the pdf
# def probability_density_function(x, y, z):
#     if x >= 0 and x <= 0.5:
#         return np.zeros(1)
#     return np.array([np.exp(-(x**2 + y**2 + z**2))])


# Set the pdf to 0 if there is an obstacle (NB: hardcoded)
def picewise_to_zero(x, y, z):
    if X_LOW < x < X_HIGH and Y_LOW < y < Y_HIGH and Z_LOW < z < Z_HIGH:
        return True
    return False


# Multivariate Gaussian PDF in R^3
def probability_density_function(x, y, z):
    # if picewise_to_zero:
    #     return np.zeros(1)

    mean = np.zeros(3)  # Mean vector
    covariance = np.eye(3)  # Covariance matrix

    point = np.array([x, y, z])
    exponent = -0.5 * np.dot(np.dot((point - mean).T, np.linalg.inv(covariance)), (point - mean))
    normalization = (2 * np.pi) ** (-len(mean) / 2) * np.linalg.det(covariance) ** (-0.5)

    return np.array([normalization * np.exp(exponent)])


# Centroid pdf
def centroid_pdf(x, y, z):
    return np.array([x,y,z]) * probability_density_function(x,y,z)


# Compute the volume of the polyhedron using convex hull.
def polyhedron_volume(vertices):
    hull = ConvexHull(vertices)
    return hull.volume


# Find Voronoi cell mass
def integrate_density_over_polyhedron(pdf_func, vertices):
    # Compute the volume of the polyhedron
    polyhedron_vol = polyhedron_volume(vertices)

    # Define the function to integrate, which is the product of the density function and the volume element
    def integrand(x, y, z):
        return pdf_func(x, y, z)

    # Define the integration limits
    limits = [(min(coord), max(coord)) for coord in zip(*vertices)]

    # Perform the triple integral over the polyhedron
    result = [quad(lambda z: quad(lambda y: quad(lambda x: integrand(x, y, z)[i], 
                     limits[0][0], 
                     limits[0][1])[0], 
                     limits[1][0], 
                     limits[1][1])[0], 
                     limits[2][0], 
                     limits[2][1])[0]
                    for i in range(len(integrand(0,0,0)))
              ]
    # Multiply the result by the volume of the polyhedron
    return np.array(result) * polyhedron_vol


def main():
    vertices = np.array([[0, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1]])
    # vertices = np.array([[1, 1, 1], [3, 0, 0], [0, 3, 0], [3, 3, 3]])
    # vertices = np.array([[0,0,0], [1,0,0], [1,0,1], [0,0,1], [0,1,0], [1,1,0], [1,1,1], [0,1,1]])

    mv = integrate_density_over_polyhedron(probability_density_function, vertices)
    print("M_v =", mv)

    cv = integrate_density_over_polyhedron(centroid_pdf, vertices) 
    print("C_v =", cv)


if __name__ == "__main__":
    main()
